Mathlib/RingTheory/MvPolynomial/EulerIdentity.lean

```diff
@@ -0,0 +1,73 @@
+/-
+Copyright (c) 2024 Junyan Xu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Junyan Xu
+-/
+import Mathlib.Algebra.MvPolynomial.PDeriv
+import Mathlib.RingTheory.MvPolynomial.Homogeneous
+
+/-!
+# Euler's homogeneous identity
+
+## Main resutls
+
+* `IsHomogeneous.sum_X_mul_pderiv`: Euler's identity for homogeneous polynomials:
+  for a multivariate homogeneous polynomial,
+  the product of each variable with the derivative with respect to that variable
+  sums up to the degree times the polynomial.
+* `IsWeightedHomogeneous.sum_weight_X_mul_pderiv`: the weighted version of Euler's identity.
+-/
+
+namespace MvPolynomial
+
+open Finsupp
+
+variable {R σ M : Type*} [CommSemiring R] {φ : MvPolynomial σ R}
+
+protected lemma IsWeightedHomogeneous.pderiv [AddCancelCommMonoid M] {w : σ → M} {n n' : M} {i : σ}
+    (h : φ.IsWeightedHomogeneous w n) (h' : n' + w i = n) :
+    (pderiv i φ).IsWeightedHomogeneous w n' := by
+  rw [← mem_weightedHomogeneousSubmodule, weightedHomogeneousSubmodule_eq_finsupp_supported,
+    Finsupp.supported_eq_span_single] at h
+  refine Submodule.span_induction ?_ ?_ (fun p q _ _ hp hq ↦ ?_) (fun r p _ h ↦ ?_) h
+  · rintro _ ⟨m, hm, rfl⟩
+    simp_rw [single_eq_monomial, pderiv_monomial, one_mul]
+    by_cases hi : m i = 0
+    · rw [hi, Nat.cast_zero, monomial_zero]; apply isWeightedHomogeneous_zero
+    convert isWeightedHomogeneous_monomial ..
+    rw [← add_right_cancel_iff (a := w i), h', ← hm, weight_sub_single_add hi]
+  · rw [map_zero]; apply isWeightedHomogeneous_zero
+  · rw [map_add]; exact hp.add hq
+  · rw [(pderiv i).map_smul]; exact (weightedHomogeneousSubmodule ..).smul_mem _ h
+
+protected lemma IsHomogeneous.pderiv {n : ℕ} {i : σ} (h : φ.IsHomogeneous n) :
+    (pderiv i φ).IsHomogeneous (n - 1) := by
+  obtain _ | n := n
+  · rw [← totalDegree_zero_iff_isHomogeneous, totalDegree_eq_zero_iff_eq_C] at h
+    rw [h, pderiv_C]; apply isHomogeneous_zero
+  · exact IsWeightedHomogeneous.pderiv h rfl
+
+variable [Fintype σ] {n : ℕ}
+
+open Finset in
+/-- Euler's identity for weighted homogeneous polynomials. -/
+theorem IsWeightedHomogeneous.sum_weight_X_mul_pderiv {w : σ → ℕ}
+    (h : φ.IsWeightedHomogeneous w n) : ∑ i : σ, w i • (X i * pderiv i φ) = n • φ := by
+  rw [← mem_weightedHomogeneousSubmodule, weightedHomogeneousSubmodule_eq_finsupp_supported,
+    supported_eq_span_single] at h
+  refine Submodule.span_induction ?_ ?_ (fun p q _ _ hp hq ↦ ?_) (fun r p _ h ↦ ?_) h
+  · rintro _ ⟨m, hm, rfl⟩
+    simp_rw [single_eq_monomial, X_mul_pderiv_monomial, smul_smul, ← sum_smul, mul_comm (w _)]
+    congr
+    rwa [Set.mem_setOf, weight_apply, sum_fintype] at hm
```

Comment:
I didn't open VS Code, so I may be wrong, but the `rwa` tells me you have two goals here. Can you use the central dot if this is the case?

Reply:
There's a linter for multiple goals now. It's [Finsupp.sum_fintype](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/BigOperators/Finsupp.html#Finsupp.sum_fintype) that creates a side goal.